src/gpuarray_elemwise.c

```diff
@@ -321,20 +323,49 @@ static int check_basic(GpuElemwise *ge, void **args, int flags,
   n = 1;
   for (j = 0; j < nd; j++) {
     p = 0;
+    read_write_arrays_found = 0;
     for (i = 0; i < ge->n; i++) {
       if (is_array(ge->args[i])) {
         v = (GpuArray *)args[i];
-        if (ge->dims[j] != v->dimensions[j]) {
+        if (ge->dims[j] == v->dimensions[j]) {
+          /* We count the number of read-write arrays found until now
+           * that are not broadcasted. */
+          if (is_rw_array(ge->args[i]))
+            ++read_write_arrays_found;
+        } else {
           if (ISCLR(flags, GE_BROADCAST)) {
             return GA_VALUE_ERROR;
           }
           /* GE_BROADCAST is set */
           if (ge->dims[j] == 1) {
+            if (read_write_arrays_found) {
+              /* There are read-write arrays before the current array,
+               * and their (j+1)th dimension equals 1, so they would be
+               * broadcasted. We don't want that. */
+              #ifdef DEBUG
+              fprintf(stderr, "\r\n(check_basic(): read-write arrays should not be broadcasted) ");
+              #endif
+              return GA_VALUE_ERROR;
+            }
+            /* There are no read-write arrays before the current array.
+             * So broadcasting can be done safely. */
             ge->dims[j] = v->dimensions[j];
+            /* We still count the current array if it's a read-write array
+             * (useless in the current implementation, but coherent). */
+            if (is_rw_array(ge->args[i]))
+              ++read_write_arrays_found;
           } else {
             if (v->dimensions[j] != 1) {
```

Comment:
I would add a check here like `|| is_output(ge->args[i])` and it should cover the case below.